keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt

```diff
@@ -247,68 +253,80 @@ class ServerGroupHandler(
       }
     }
 
-  private suspend fun CloudDriverService.getServerGroup(resource: Resource<ServerGroupSpec>): ServerGroup? {
-    try {
-      return activeServerGroup(
-        resource.serviceAccount,
-        resource.spec.moniker.app,
-        resource.spec.location.accountName,
-        resource.spec.moniker.name,
-        resource.spec.location.region,
-        CLOUD_PROVIDER
-      )
-        .run {
-          publisher.publishEvent(ArtifactAlreadyDeployedEvent(
-            resourceId = resource.id.value,
-            imageId = launchConfig.imageId
-          ))
-
-          ServerGroup(
-            moniker = Moniker(app = moniker.app, stack = moniker.stack, detail = moniker.detail),
-            location = Location(
-              accountName,
-              region,
-              subnet,
-              zones
-            ),
-            launchConfiguration = launchConfig.run {
-              LaunchConfiguration(
-                imageId = imageId,
-                appVersion = image.appVersion,
-                instanceType = instanceType,
-                ebsOptimized = ebsOptimized,
-                iamRole = iamInstanceProfile,
-                keyPair = keyName,
-                instanceMonitoring = instanceMonitoring.enabled,
-                ramdiskId = ramdiskId.orNull()
-              )
-            },
-            capacity = capacity.let { Capacity(it.min, it.max, it.desired) },
-            dependencies = Dependencies(
-              loadBalancerNames = loadBalancers,
-              securityGroupNames = securityGroupNames,
-              targetGroups = targetGroups
-            ),
-            health = Health(
-              enabledMetrics = asg.enabledMetrics.map { Metric.valueOf(it) }.toSet(),
-              cooldown = asg.defaultCooldown.let(Duration::ofSeconds),
-              warmup = asg.healthCheckGracePeriod.let(Duration::ofSeconds),
-              healthCheckType = asg.healthCheckType.let { HealthCheckType.valueOf(it) },
-              terminationPolicies = asg.terminationPolicies.map { TerminationPolicy.valueOf(it) }.toSet()
-            ),
-            scaling = Scaling(
-              suspendedProcesses = asg.suspendedProcesses.map { ScalingProcess.valueOf(it) }.toSet()
-            ),
-            tags = asg.tags.associateBy(Tag::key, Tag::value).filterNot { it.key in DEFAULT_TAGS }
-          )
+  private suspend fun CloudDriverService.getServerGroups(resource: Resource<ClusterSpec>): Set<ServerGroup> =
+    coroutineScope {
+      resource.spec.locations.regions.map {
+        async {
+          try {
+            activeServerGroup(
```

Comment:
For simplicity I've currently made this code use the same River endpoint as before so we're making _n_ calls to River per cluster. There may be a way to optimize this later but I consider that out of scope for this PR.